co_return to accept same expressions and types as return

bashtanov opened the pull request:
Current implementation of `co_return` behaves differently from plain
`return` when interpreting the expression returned. This includes both
- `co_return` accepting expressions `return` would reject;
- `co_return` rejecting expressions `return` would accept.

Unexpectedly accepted expressions allow bugs from accidental misuse:
- a value of a strongly-defined type may be implicitly instantiated from
its underlying type thus defeating the purpose, e.g. a coroutine
declared to return `future<std::chrono::seconds>`, may `co_return 42;`
- instead of a value of an enum type used in the future, a value of an
unrelated enum type may be used if it uses the same underlying type;
- obscure convertions, such as a number to `std::vector`, may be applied
implicitly.

Unexpectedly rejected expressions include:
- those rejected due to false ambiguity, where compiler cannot decide
between a reasonable interpretation and one of the situtaions described
above;
- brace-enclosed initializer list without type specified.

This happens because we save the `co_return`'ed value in a state by:
- calling a chain of unnecessarily templated functions;
- implicitly calling the type ctor even if it is marked explicit.

The fix is to avoid the above.

One of the test cases was shamelessly stolen from https://github.com/scylladb/seastar/pull/2971 as my change fixes their use case.

bashtanov committed:
core/{coroutine,future}: preparations to fix `co_return` type semantics

Backwards compatible so far, so no API version bump.
1. Split `promise` and `future` internal functions for setting stored
values: separate those emplace-like and those only working with an
argument of an implicitly convertible type.
2. Make `coroutine_traits_base::return_value` accept only one argument,
as that's how `co_await` invokes it.
3. Add `promise::emplace_value` for constructing a value in place
including using an explcit constructor. Strict emplace semantics, as
opposed to `set_value`, which has special rules for tuples.
4. Internally use set/emplace where appropriate.

bashtanov committed:
tests/unit/{coroutines,futures}: strict types on co_return and set_value

- co_return interprets expressions like return does
- promise.set_value employs only implicit conversions
- make_ready_future can both construct a value and pass an existing one
- no excessive moves or copies happen

bashtanov committed:
api: introduce version 10:

`promise::set_value` and `co_return` to use the same semantics as plain
return. `promise::emplace_value` can be used for in-place construction.
`make_ready_future` to support both, but not expanding tuples.

bashtanov committed:
core/{coroutine,future}: make `co_return` more strict with types

Current implementation of `co_return` behaves differently from plain
`return` when interpreting the expression returned. This includes both
- `co_return` accepting expressions `return` would reject;
- `co_return` rejecting expressions `return` would accept.

Unexpectedly accepted expressions allow bugs from accidental misuse:
- a value of a strongly-defined type may be implicitly instantiated from
its underlying type thus defeating the purpose, e.g. a coroutine
declared to return `future<std::chrono::seconds>`, may `co_return 42;`
- instead of a value of an enum type used in the future, a value of an
unrelated enum type may be used if it uses the same underlying type;
- obscure convertions, such as a number to `std::vector`, may be applied
implicitly.

Unexpectedly rejected expressions include:
- those rejected due to false ambiguity, where compiler cannot decide
between a reasonable interpretation and one of the situtaions described
above;
- brace-enclosed initializer list without type specified.

This happens because we save the `co_return`'ed value in a state by:
- calling a chain of unnecessarily templated functions;
- implicitly calling the type ctor even if it is marked explicit.

The fix is to only accept values of types std::is_convertible to the
declared type.

The changes are only to affect API version 10, which is to be enabled
in a subsequent commit.

avikivity commented:
If you update the series, kindly inform the reviewer about what has changed rather than having them reverse-engineer it.

bashtanov commented:
@avikivity sure, there we go:
- To avoid excessive moves, implementation changed:
   - the argument is passed via the function chain perfectly forwarded, but every time with a static check that it is implicitly convertible to the target type
   - the actual conversion happens when the value is stored by calling a return-its-argument function `implicit_convert`, copy/move elided thanks to C++17 rules
- As requested, split commits into preparations and user-facing header changes
- Added more tests, mostly to make sure no excessive moves or copies happen
- Rebased
- Bumped supported C++ versions to 23 and 26 in docs. I used `new (placement) auto (...)` syntax I thought was C++23-specific, but now I double checked and it seems that, unlike function-style cast `auto(foo)`, it's been supported since C++11. I'll remove it in the next review iteration if needed.

Hope it helps. If any code or commit will benefit from a more detailed comment I would appreciate if you could let me know.

avikivity commented on include/seastar/core/coroutine.hh:
```diff
@@ -95,9 +95,14 @@ public:
+
+        [[deprecated("Forwarding coroutine returns are deprecated as too dangerous. Use 'co_return co_await ...' until explicit syntax is available.")]]
+        void return_value(future<T>&& fut) noexcept {
+            fut.forward_to(std::move(_promise));
```
Why this code movement?

bashtanov replied:
to group functions that only make sense for API_LEVEL <= 9

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -290,16 +297,33 @@ struct uninitialized_wrapper {
     uninitialized_wrapper() noexcept = default;
+    template<typename U>
+    requires std::is_convertible_v<U&&, T>
+    void uninitialized_set(U&& u) {
```
Slightly better to use std::convertible_to from the concept library (an even say `uninialized_set(auto std::convertible_to<T>&& u)`

bashtanov replied:
1) `std::convertible_to<From, To>` is semantically different: it checks that `From` can be converted to `To` _both_ implicitly and explicitly. Is there a pair of types convertible implicitly but not explicitly? [Yes!](https://stackoverflow.com/a/62644127) If a type is not return'able it shouldn't be co_return'able.
2) The concept syntactic sugar is still achievable if we create a concept based on `std::is_convertible_v`. However, the desired condition here is `std::is_convertible_v<U&&, T>`, not `std::is_convertible_v<U, T>`. Assuming `U&&` is well-formed, these conditions are actually equivalent, see proof below. And `U&&` is present as a function argument. Would you like me to implement such concept and use it here and in other places instead of `requires std::is_convertible_v<U&&, T>`?
----
Proof:
`std::is_convertible_v<From, To>` tests whether `To f() { return std::declval<From>(); }` is well-formed. And `std::declval<T>()` returns `std::add_rvalue_reference_t<T>`. So:
`is_convertible_v<U, T>` effectively tests `add_rvalue_reference_t<U> → T`
`is_convertible_v<U&&, T>` effectively tests `add_rvalue_reference_t<U&&> → T`
By reference collapsing, these produce identical types in every well-formed case.

avikivity replied:
The main advantage of the library concept is that it's a library concept. Not a random concept that people have to study (and get lost in).

Let's keep it as is.

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -290,16 +297,33 @@ struct uninitialized_wrapper {
     requires (!std::same_as<std::tuple<std::remove_cv_t<U>...>, std::tuple<tuple_type>>)
-    void
-    uninitialized_set(U&&... vs) {
+    void uninitialized_set_legacy(U&&... vs) {
```
"legacy" doesn't mean much outside the point in time it is introduced.

Try to find a descriptive name (unless it's gone by the end of the series).

bashtanov replied:
Any preference between `_loose`, `_lax`, `_relaxed`, `_compat`? My fav is `_lax`.

avikivity replied:
Let's be descriptive, loose_type_conversion.

btw, do we not have the same problem in future::set_value? and in make_ready_future? If we do this surgery, it's better to capture all the problematic areas at once.

bashtanov replied:
The naming problem? I'll change it in all places I introduced the `_legacy` suffix.

Or the general type conversion problem?
For `promise::set_value` I changed the semantics of `set_value` and introduced `emplace_value`.
As for `make_ready_future`, I kept it loose, but made sure the set semantics is honoured -- for omitting types accepting braced lists.

I'll add tests for these to illustrate.

avikivity replied:
We should consider having emplace_ready_future vs make_ready_future.

Not sure what the prior art is around here (in the Standard).

bashtanov replied:
> We should consider having emplace_ready_future vs make_ready_future.

We have them in a sense:
1) `make_ready_future<specified_type>(whatever, arguments)` has emplace semantics.
2) `make_ready_future(single_argument)` has copy/move semantics.
That's similar to the logic of `make_lw_shared` and `std::make_optional`.

Unlike these, `co_return` and `promise::set_value` have prior type info from the object, so they are typically called without explicit template parameters.

The only thing I need change for make_ready_future is to make it 100% compatible with the pre-change version for API_VERSION <=9.

> Not sure what the prior art is around here (in the Standard).

Could you rephrase? I'm not sure I understand it right.

avikivity replied:
We have vector::push_back(T) that allows the implicit conversion during argument construction, and vector::emplace_back(A&&...) that explicitly constructs a T from the tuple of A&&s. That's prior are. It matches well for the proposed promise::set_value() and promise::emplace_value().

I'm asking if we can draw a similar parallel for make_ready_future: one name for the implicit-only conversion, and one for explicit construct-from-tuple. I don't think the name emplace_ready_future() works, because we're not emplacing it anywhere. Practically all of the use is returning the same type as the single input, I think it makes sense to have a special name for the constructing variant.

bashtanov replied:
`vector<T>` is similar to `promise<T>` and `future<T>`: the object carries the type info, so neither `push_back` nor `emplace_back` need to specify `T` again. Let me illustrate why it matters.

---
We have 4 ways to set a value for a promise `pr` of type `promise<T>`:
1) `pr.set_value(a value of type T)` -- put as is
2) `pr.set_value(something implicitly convertible to T)` -- implicitly convert and put
3) `pr.emplace_value(any parameters)` -- invoke a ctor, possibly an explicit one; in case of a single argument explicit ctors are called
4) `pr.set_value(T(something))` -- invoke an explicit conversion (whether a ctor or not)

The main reason to have different function names (set vs emplace) is to distinguish between situations (2) and (3): set_value, like vector's push_back, is not expected to invoke an explicit ctor, not least because T is not mentioned in the call (2).

---
In case of `make_ready_future` (or, similarly, `make_lw_shared_ptr` or `make_optional`) we don't yet have an object that holds type info. We have the following ways to call it:
1) `make_ready_future(a value of type T)` -- put as is; this includes supplying a copyable lvalue or a reference -- we still put the value as we consider `future<Foo&>` marginal
2) `make_ready_future<T>(something else implicitly convertible to T)` -- invoke an implicit convertion; note that we need to specity T in angle brackets to define the result type
3) `make_ready_future<T>(any parameters)` -- invoke a ctor, possibly an explicit one; in case of a single argument explicit ctors are called but not other explicit conversions
4) `make_ready_future(T(something))` -- invoke an explicit conversion; note that here it's not longer to type than (2)

Here we don't mind the syntax clash between implicit (2) and potentially-calling-explicit-ctor (3), as they both mention T. There's no risk of accidentally invoking an explicit conversion without ever mentioning the target type -- the very situation this PR fights against.

---
A direct analogy would be `std::optional`. It uses separate `emplace` and `operator=(T&& value )` for changing the value of an existing object, but a single `make_optional` for creating a new one. They cover all the 4 situations above.

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -290,16 +297,33 @@ struct uninitialized_wrapper {
+    requires std::is_convertible_v<U&&, T>
+    void uninitialized_set(U&& u) {
+        // Despite the `requires` condition above, the line below may still fail compilation if T is a reference type.
+        // E.g. when T is a const lvalue reference and U is an rvalue reference to a non-const type.
```
Do we even support future\<T&>?

bashtanov replied:
[We do](https://github.com/scylladb/seastar/blob/master/include/seastar/core/future.hh#L287-L288)

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -612,18 +650,59 @@ struct future_state :  public future_state_base, private internal::uninitialized
+    template<typename U>
+    requires std::is_convertible_v<U&&, T>
+    future_state(set_ready_future_marker, U&& v) noexcept : future_state_base(state::result) {
+      try {
```
We use 4-space indents.

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -612,18 +650,59 @@ struct future_state :  public future_state_base, private internal::uninitialized
+      try {
+        this->uninitialized_set(std::forward<U>(v));
+      } catch (...) {
+        new (this) future_state(current_exception_future_marker());
```
I don't know if it's legal to double-construct *this.

bashtanov replied:
This comes from existing code too. I don't like it either, it does something nasty about object lifetimes. The standard is vague about whether it's legal. But fixing it is probably out of scope.

avikivity replied:
Ok. Existing code gets a pass.

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -612,18 +650,59 @@ struct future_state :  public future_state_base, private internal::uninitialized
+    template <typename... A>
+    void set_legacy(A&&... a) noexcept {
+        assert(_u.st == state::future);
+        new (this) future_state(legacy_ready_future_marker(), std::forward<A>(a)...);
```
Again I don't like double construction.  Since this is from existing code I'll allow it here.

bashtanov replied:
This is legal due to [[basic.life]/6](https://eel.is/c++draft/basic.life#6)

avikivity commented on include/seastar/core/coroutine.hh:
```diff
@@ -104,6 +105,18 @@ public:
+        }
+
+        template<typename U>
+        requires std::is_convertible_v<U&&, T>
```
Again prefer std::convertible_to

avikivity commented on tests/unit/coroutines_test.cc:
```diff
@@ -946,3 +949,186 @@ SEASTAR_TEST_CASE(test_try_future) {
+
+SEASTAR_TEST_CASE(test_std_expected_void_specialization) {
+  BOOST_REQUIRE_EQUAL(co_await co_return_tup_int_rv(), 42);
+  BOOST_REQUIRE_EQUAL(co_await co_return_tup_int_clv(), 42);
```
Again 4-space indents.

avikivity commented on README.md:
```diff
@@ -146,11 +146,11 @@ There are also instructions for building on any host that supports [Docker](doc/
 Use of the [DPDK](http://dpdk.org) is [optional](doc/building-dpdk.md).
 
-#### Seastar's C++ standard: C++20 or C++23
+#### Seastar's C++ standard: C++23 or C++26
```
This doesn't belong in this series.

bashtanov commented:
changes:
- renamed *_legacy as discussed
- removed C++ version change from readme
- fixed formatting issues
- added tests for promise.set_value and make_ready_future

bashtanov commented:
changes:
* made `make_ready_future` 100% compatible with the pre-change version for API_VERSION <=9
* added a test for it
* removed an incorrect code comment

avikivity commented on include/seastar/core/coroutine.hh:
```diff
@@ -95,9 +95,14 @@ public:
-            _promise.set_value(std::forward<U>(value)...);
+        template<typename U>
+        void return_value(U&& value) {
+            _promise.set_value(std::forward<U>(value));
```
I suppose this is a safe change because there is no way to get this called with sizeof...(U) != 1?

bashtanov replied:
yes, `return_value` is only called by co_return with a single argument https://eel.is/c++draft/stmt.return.coroutine#2.1

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -290,16 +297,33 @@ struct uninitialized_wrapper {
+        // Despite the `requires` condition above, the line below may still fail compilation if T is a reference type.
+        // E.g. when T is a const lvalue reference and U is an rvalue reference to a non-const type.
+        // These kind of conversions make little sense both here and with plain `return`.
+        new(&_v.value) auto(implicit_convert<maybe_wrap_ref<T>>(std::forward<U>(u)));
```
IIUC auto(x) is C++23 only while we are still C++20/23.

Maybe we can push the C++23/26 patch before this series (but outside it) so we don't have to change this.

bashtanov replied:
auto(x) is C++23, but new auto(x) is C++11

avikivity replied:
Ah, thanks.

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -290,16 +297,33 @@ struct uninitialized_wrapper {
         uninitialized_set(std::move(std::get<0>(v)));
     }
-    void uninitialized_set(const tuple_type& v) {
+    void uninitialized_set_legacy(const tuple_type& v) {
```
Here is legacy again

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -557,6 +591,10 @@ inline void future_state_base::any::check_failure() noexcept {
+struct loose_type_conversion_ready_future_marker: ready_future_marker {};
+struct set_ready_future_marker: ready_future_marker {};
+struct set_from_tuple_ready_future_marker: ready_future_marker {};
+struct emplace_ready_future_marker: ready_future_marker {};
```
Why the inheritance?

Style: we have space before :.

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -1232,8 +1355,8 @@ private:
-    template <typename... A>
-    future(ready_future_marker m, A&&... a) noexcept : _state(m, std::forward<A>(a)...) { }
+    template <std::derived_from<ready_future_marker> Marker, typename... A>
+    future(Marker m, A&&... a) noexcept : _state(m, std::forward<A>(a)...) { }
```
Okay, I see the reason for the inheritance. I think it's better to be explicit and just repeat the constructor for every marker, than risk decaying into the plain marker and being surprised.

avikivity commented on doc/compatibility.md:
```diff
@@ -78,6 +78,10 @@ versions of the API. For example.
+   - Seastar_API_LEVEL=10 makes co_return and promise.set_value() semantics
+     closer to those of plain C++ return regarding how they convert
+     expressions used as their arguments to the declared type.
+
```
Please extend the history table below.

avikivity commented on tests/unit/CMakeLists.txt:
```diff
@@ -346,7 +346,9 @@ seastar_add_test (content_source
-  SOURCES coroutines_test.cc)
+  SOURCES
+    coroutines_test.cc
+    conversion_test_types.hh)
```
Why do you need to add .hh? Won't the build system deduce it?

avikivity commented:
Looks good, I added low-level comments. Main remaining point is whether we want to split make_ready_future into two names. If we do, I think it makes sense to do it in this series as it's the same problem.

bashtanov commented:
changes:
- fixed build errors
- removed `ready_future_marker` base class
- added v10 to API version table
- removed conversion_test_types.hh from CMakeLists.txt

bashtanov commented:
The failure seems to be an infrastructure issue
